# load required packages
library(Signac);library(Seurat);library(ensembldb);library(EnsDb.Mmusculus.v79);library(patchwork);library(RColorBrewer);library(stringr)
library(ChIPseeker);library("TxDb.Mmusculus.UCSC.mm10.knownGene");library("org.Mm.eg.db");library(BSgenome.Mmusculus.UCSC.mm10);
library(EnsDb.Mmusculus.v79);library(JASPAR2020);library(TFBSTools);library(motifmatchr)

set.seed(1234)

# read dataset from outs folder of cellranger count
counts <- Read10X_h5("~/multi/filtered_feature_bc_matrix.h5")
# Initialize the Seurat object for GEX library alone
multi.gex <- CreateSeuratObject(
  counts = counts$`Gene Expression`,
  assay = "RNA"
)
multi.gex[["percent.mt"]] <- PercentageFeatureSet(multi.gex, pattern = "^mt-")
multi.gex <- PercentageFeatureSet(multi.gex, pattern = "^mt-", col.name = "percent.mt")

# Visualize QC metrics as a violin plot
VlnPlot(multi.gex, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# Setup the Seurat objects for unsupervised clustering analysis
multi.gex <- subset(multi.gex, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & percent.mt < 20)
multi.gex <- SCTransform(multi.gex, vars.to.regress = "percent.mt", verbose = FALSE)
multi.gex <- RunPCA(multi.gex, verbose = FALSE)
multi.gex <- RunUMAP(multi.gex, dims = 1:30, verbose = FALSE)
multi.gex <- FindNeighbors(multi.gex, dims = 1:30, verbose = FALSE)
multi.gex <- FindClusters(multi.gex, verbose = FALSE)
DimPlot(multi.gex, label = TRUE) + NoLegend()

# Initialize the Seurat object for both GEX and ATAC library

fragpath <- "~/multi/atac_fragments.tsv.gz"

# get gene annotations for mm10
edb <- EnsDb.Mmusculus.v79
seqlevelsStyle(edb) <- "UCSC"
annotations <- GetGRangesFromEnsDb(ensdb = edb)
genome(annotations) <- "mm10"

# create a Seurat object containing the RNA adata
multi.joint <- CreateSeuratObject(
  counts = counts$`Gene Expression`,
  assay = "RNA"
)

# create ATAC assay and add it to the object
multi.joint[["ATAC"]] <- CreateChromatinAssay(
  counts = counts$Peaks,
  sep = c(":", "-"),
  fragments = fragpath,
  annotation = annotations
)

DefaultAssay(multi.joint) <- "ATAC"

multi.joint <- NucleosomeSignal(multi.joint)
multi.joint <- TSSEnrichment(multi.joint)

VlnPlot(
  object = multi.joint,
  features = c("nCount_RNA", "nCount_ATAC", "TSS.enrichment", "nucleosome_signal"),
  ncol = 4,
  pt.size = 0
)

# filter out low quality cells
multi.joint <- subset(
  x = multi.joint,
  subset = nCount_ATAC < 100000 &
    nCount_RNA < 40000 &
    nCount_ATAC > 1000 &
    nCount_RNA > 1000 &
    nucleosome_signal < 1 &
    TSS.enrichment > 1
)


# Data processing
DefaultAssay(multi.joint) <- "RNA"
multi.joint <- SCTransform(multi.joint)
multi.joint <- RunPCA(multi.joint)

DefaultAssay(multi.joint) <- "ATAC"
multi.joint <- FindTopFeatures(multi.joint, min.cutoff = 5)
multi.joint <- RunTFIDF(multi.joint)
multi.joint <- RunSVD(multi.joint)

# build a joint neighbor graph using both assays
multi.joint <- FindMultiModalNeighbors(
  object = multi.joint,
  reduction.list = list("pca", "lsi"), 
  dims.list = list(1:50, 2:40),
  modality.weight.name = "RNA.weight",
  verbose = TRUE
)

# build a joint UMAP visualization
multi.joint <- RunUMAP(
  object = multi.joint,
  nn.name = "weighted.nn",
  assay = "RNA",
  verbose = TRUE
)
DimPlot(multi.joint, label = TRUE, repel = TRUE, reduction = "umap") + NoLegend()

# Generation of Heatmap
gex.markers <- FindAllMarkers(multi.temp, only.pos = TRUE, min.pct = 0.5, logfc.threshold = 0.5)
top10 <- gex.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(subset(multi.temp,downsample = 400 ), features = top10$gene,group.by = 'ident',draw.lines = T,lines.width = 10,label = F) + NoLegend()

# Generation of Donutplot
donut <- table(Idents(multi.joint),multi.joint$orig.ident)
donut <- as.data.frame(donut)
library(ggpubr)
ggdonutchart(donut,'cells',fill = 'cluster',color = 'white') + theme(legend.position = "right",legend.title=element_blank(),legend.text=element_text(size=20),axis.text.x =element_text(size=0))

library(ChIPseeker)
library("TxDb.Mmusculus.UCSC.mm10.knownGene")
library("org.Mm.eg.db")

# Generation peak list from each group

for(i in levels(multi.joint)) {
  peak.list <- list()
  peak.list[[i]] <- AccessiblePeaks(multi.joiny, idents = i)
  peak.list[[i]] <- StringToGRanges(peak.list[[i]], sep = c("-", "-"))
  assign(paste("cell",".",i,sep = ""), peak.list[[i]])
  
  peak.list <- GenomicRanges::GRangesList(`CAP1 (gCAP)` = readPeakFile(`cell.CAP1 (gCAP)`),
                                        `CAP2 (aCAP)` = readPeakFile(`cell.CAP2 (aCAP)`),
                                        Artery = readPeakFile(cell.Artery),
                                        Venous = readPeakFile(cell.Venous),
                                        Lymphatics = readPeakFile(cell.Lymphatics),
                                        Pericyte = readPeakFile(cell.Pericyte),
                                        Fibroblast = readPeakFile(cell.Fibroblast),
                                        Myofibroblast = readPeakFile(cell.Myofibroblast),
                                        Matrixfibroblast = readPeakFile(cell.Matrixfibroblast))
library(scales)
color.use <- as.vector(hue_pal()(n=9))                                     
peakHeatmap(peak.list, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, upstream=3000, downstream=3000, color=color.use)                                        




# Generation of Dotplot
DotPlot(acd.combined, features = c('Acvrl1','Acvr1b','Acvr1c','Bmpr1a','Bmpr1b','Tgfbr1','Tgfbr2','Acvr2a','Acvr2b','Bmpr2'), cols = c("blue","red"), dot.scale = 8,split.by = "orig.ident") + RotatedAxis() + coord_flip()
# Generation of correlation plot among clusters
palette = colorRampPalette(c("green", "white", "red")) (20)
av.exp <- AverageExpression(acd.combined)$RNA
cor.exp <- as.matrix(cor(av.exp))
heatmap(x = cor.exp, col = palette,Rowv = NA,symm = TRUE,scale = "row",margins = c(8, 8))
# Cells distribution across clusters and pie plot
cell.prop <- prop.table(table(Idents(acd.combined),acd.combined$orig.ident),2)
cell.prop <- 100*cell.prop
cell.prop <- as.data.frame(cell.prop)
names(cell.prop)[1:3] <- c('Cluster','group','value')
ggplot(cell.prop, aes(fill=Cluster, y=value, x=group)) + geom_bar(position="stack", stat="identity") + theme_bw() + scale_y_continuous(expand = c(0,0),limits = c (0,100)) + theme(legend.position = "right",legend.title=element_blank(),legend.text=element_text(size=15),axis.title.x =element_text(size=0),axis.title.y=element_text(size=0),axis.text.x =element_text(size=0),axis.text.y=element_text(size=15))
ggplot(cell.prop[1:12,], aes(fill=Cluster, y=value, x='')) + geom_bar(stat="identity", width = 1) + coord_polar("y", start=0) + theme_void() + NoLegend()
ggplot(cell.prop[13:24,], aes(fill=Cluster, y=value, x='')) + geom_bar(stat="identity", width = 1) + coord_polar("y", start=0) + theme_void() + NoLegend()
# 





