# load required packages
library(Signac);library(Seurat);library(ensembldb);library(EnsDb.Mmusculus.v79);library(patchwork);library(RColorBrewer);library(stringr)
library(ChIPseeker);library("TxDb.Mmusculus.UCSC.mm10.knownGene");library("org.Mm.eg.db");library(BSgenome.Mmusculus.UCSC.mm10);
library(EnsDb.Mmusculus.v79);library(JASPAR2020);library(TFBSTools);library(motifmatchr)

set.seed(1234)

# read dataset from outs folder of cellranger count
counts <- Read10X_h5("~/multi/filtered_feature_bc_matrix.h5")
# Initialize the Seurat object for GEX library alone
multi.gex <- CreateSeuratObject(
  counts = counts$`Gene Expression`,
  assay = "RNA"
)
multi.gex[["percent.mt"]] <- PercentageFeatureSet(multi.gex, pattern = "^mt-")
multi.gex <- PercentageFeatureSet(multi.gex, pattern = "^mt-", col.name = "percent.mt")

# Visualize QC metrics as a violin plot
VlnPlot(multi.gex, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# Setup the Seurat objects for unsupervised clustering analysis
multi.gex <- subset(multi.gex, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & percent.mt < 20)
multi.gex <- SCTransform(multi.gex, vars.to.regress = "percent.mt", verbose = FALSE)
multi.gex <- RunPCA(multi.gex, verbose = FALSE)
multi.gex <- RunUMAP(multi.gex, dims = 1:30, verbose = FALSE)
multi.gex <- FindNeighbors(multi.gex, dims = 1:30, verbose = FALSE)
multi.gex <- FindClusters(multi.gex, verbose = FALSE)
DimPlot(multi.gex, label = TRUE) + NoLegend()

# Initialize the Seurat object for both GEX and ATAC library

fragpath <- "~/multi/atac_fragments.tsv.gz"

# get gene annotations for mm10
edb <- EnsDb.Mmusculus.v79
seqlevelsStyle(edb) <- "UCSC"
annotations <- GetGRangesFromEnsDb(ensdb = edb)
genome(annotations) <- "mm10"

# create a Seurat object containing the RNA adata
multi.joint <- CreateSeuratObject(
  counts = counts$`Gene Expression`,
  assay = "RNA"
)

# create ATAC assay and add it to the object
multi.joint[["ATAC"]] <- CreateChromatinAssay(
  counts = counts$Peaks,
  sep = c(":", "-"),
  fragments = fragpath,
  annotation = annotations
)

DefaultAssay(multi.joint) <- "ATAC"

multi.joint <- NucleosomeSignal(multi.joint)
multi.joint <- TSSEnrichment(multi.joint)

VlnPlot(
  object = multi.joint,
  features = c("nCount_RNA", "nCount_ATAC", "TSS.enrichment", "nucleosome_signal"),
  ncol = 4,
  pt.size = 0
)

# filter out low quality cells
multi.joint <- subset(
  x = multi.joint,
  subset = nCount_ATAC < 100000 &
    nCount_RNA < 40000 &
    nCount_ATAC > 1000 &
    nCount_RNA > 1000 &
    nucleosome_signal < 1 &
    TSS.enrichment > 1
)


# Data processing
DefaultAssay(multi.joint) <- "RNA"
multi.joint <- SCTransform(multi.joint)
multi.joint <- RunPCA(multi.joint)

DefaultAssay(multi.joint) <- "ATAC"
multi.joint <- FindTopFeatures(multi.joint, min.cutoff = 5)
multi.joint <- RunTFIDF(multi.joint)
multi.joint <- RunSVD(multi.joint)

# build a joint neighbor graph using both assays
multi.joint <- FindMultiModalNeighbors(
  object = multi.joint,
  reduction.list = list("pca", "lsi"), 
  dims.list = list(1:50, 2:40),
  modality.weight.name = "RNA.weight",
  verbose = TRUE
)

# build a joint UMAP visualization
multi.joint <- RunUMAP(
  object = multi.joint,
  nn.name = "weighted.nn",
  assay = "RNA",
  verbose = TRUE
)
DimPlot(multi.joint, label = TRUE, repel = TRUE, reduction = "umap") + NoLegend()

# Generation of Heatmap
gex.markers <- FindAllMarkers(multi.temp, only.pos = TRUE, min.pct = 0.5, logfc.threshold = 0.5)
top10 <- gex.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(subset(multi.temp,downsample = 400 ), features = top10$gene,group.by = 'ident',draw.lines = T,lines.width = 10,label = F) + NoLegend()

# Generation of Donutplot
donut <- table(Idents(multi.joint),multi.joint$orig.ident)
donut <- as.data.frame(donut)
donut$cluster <- factor(donut$cluster,c("CAP1 (gCAP)", "CAP2 (aCAP)", "Artery","Venous","Lymphatics","Pericyte","Fibroblast","Myofibroblast" ,"Matrixfibroblast"))
library(ggpubr)
ggdonutchart(donut,'cells',fill = 'cluster',color = 'white') + theme(legend.position = "right",legend.title=element_blank(),legend.text=element_text(size=20),axis.text.x =element_text(size=0))

# Generation distance plot of TSS from each cell cluster
library(ChIPseeker)
library("TxDb.Mmusculus.UCSC.mm10.knownGene")
library("org.Mm.eg.db")
annopeakfile <- lapply(peak.list, annotatePeak, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,tssRegion=c(-3000, 3000),annoDb="org.Mm.eg.db")
plotDistToTSS(annopeakfile) + theme(legend.position = "right",legend.title=element_blank(),legend.text=element_text(size=20),title =element_text(size=0),axis.title.x =element_text(size=20),axis.title.y=element_text(size=20),axis.text.x =element_text(size=20),axis.text.y=element_text(size=20))
                               
# Generation of Foxf1 Footprint
multi.motif <- Footprint(object = multi.motif,motif.name = "Foxf1",genome = BSgenome.Mmusculus.UCSC.mm10)
PlotFootprint(multi.motif, features = "Foxf1")

# Get a list of motif position frequency matrices from the JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)

# add motif information
multi.joint <- AddMotifs(
  object = multi.joint,
  genome = BSgenome.Mmusculus.UCSC.mm10,
  pfm = pfm
)

multi.joint <- RunChromVAR(
  object = multi.joint,
  genome = BSgenome.Mmusculus.UCSC.mm10
)


DefaultAssay(multi.joint) <- 'chromvar'

# look at the activity of Foxf1
FeaturePlot(object = multi.joint,features = "MA1606.1",min.cutoff = 'q10',max.cutoff = 'q90',pt.size = 0.1)

# find cell type TF enrichment and unique
tf.enrich <- FindAllMarkers(multi.motif, only.pos = T, assay = "chromvar")
top30 <- tf.enrich %>% group_by(cluster) %>% top_n(n = 30, wt = avg_log2FC)
tf.unique <- unique(tf.enrich$gene)
chrommat <- AverageExpression(multi.motif, assays = "chromvar", features = tf.unique) %>%
  as.data.frame()
tf.uniquename <- ConvertMotifID(object = multi.motif, id = tf.unique)
names(chrommat) <- levels(multi.motif)
rownames(chrommat) <- tf.uniquename

# sort the matrix to find column index for TF activity
chromemat <- chromemat[do.call(order, c(chromemat, list(decreasing=TRUE))),]
chromemat$max <- max.col(chromemat)
chromemat <- chromemat[order(chromemat$max), ]
chromemat <- dplyr::select(chromemat, -max)

ann_colors = list(cluster = c(`CAP1 (gCAP)` = '#F8766D', `CAP2 (aCAP)` = '#D39200', Artery = '#93AA00',Venous = '#00BA38', Lymphatics = '#00C19F', Pericyte = '#00B9E3',Fibroblast = '#619CFF', Myofibroblast = '#DB72FB',Matrixfibroblast = '#FF61C3'))
color.list <- setNames(split(color.data, seq(nrow(color.data))), rownames(color.data))
pheatmap(chromemat,scale = "row",cluster_cols=F,cluster_rows = F,annotation_col = col.data, annotation_colors = color.list,angle_col = 315,color = colorRampPalette(c("blue","white","red"))(100),show_rownames=F)

# Generation of Vlnplot of Foxf1 expression
VlnPlot(multi.gex, features = c("Foxf1", "Ebf1"), ncol = 2,pt.size = 0)

# Coverageplot of Foxf1 genomic region
bw.path <- "~/bigwig/"
CoveragePlot(
object = multi.joint,
region = "chr8-120800001-121100000",
annotation = T,
bigwig = bw.path,
bigwig.type = "coverage"
ncol = 1
)
# plot regions of human genome for both donor and patient
CoveragePlot(
  object = acd5.joint,
  region = "chr16-85920001-86520000",
  annotation = T,
  peaks = T,
  ncol = 1
)

CoveragePlot(
  object = acd5.joint,
  region = "chr16-86020001-86040001",
  annotation = F,
  peaks = F,
  ncol = 1
)

CoveragePlot(
  object = acd5.joint,
  region = "chr16-86140001-86160001",
  annotation = F,
  peaks = F,
  ncol = 1
)

CoveragePlot(
  object = acd5.joint,
  region = "chr16-86075001-86095001",
  annotation = F,
  peaks = F,
  ncol = 1
)
CoveragePlot(
  object = acd5.joint,
  region = "chr16-86175001-86195001",
  annotation = F,
  peaks = F,
  ncol = 1
)
CoveragePlot(
  object = acd5.joint,
  region = "chr16-86500001-86520001",
  annotation = F,
  peaks = F,
  ncol = 1
)
